pkg/skaffold/runner/runner.go

```diff
@@ -128,25 +132,97 @@ func NewForConfig(opts *config.SkaffoldOptions, cfg *latest.SkaffoldConfig) (*Sk
 
 	monitor := filemon.NewMonitor()
 
-	return &SkaffoldRunner{
+	intentChan := make(chan bool, 1)
+
+	r := &SkaffoldRunner{
 		Builder:  builder,
 		Tester:   tester,
 		Deployer: deployer,
 		Tagger:   tagger,
 		Syncer:   kubectl.NewSyncer(runCtx.Namespaces),
 		monitor:  monitor,
 		listener: &SkaffoldListener{
-			Monitor: monitor,
-			Trigger: trigger,
+			Monitor:    monitor,
+			Trigger:    trigger,
+			intentChan: intentChan,
+		},
+		changeSet: &changeSet{
+			rebuildTracker: make(map[string]*latest.Artifact),
+			resyncTracker:  make(map[string]*sync.Item),
 		},
-		changeSet:            &changeSet{},
 		labellers:            labellers,
 		defaultLabeller:      defaultLabeller,
 		portForwardResources: cfg.PortForward,
 		imageList:            kubernetes.NewImageList(),
 		cache:                artifactCache,
 		runCtx:               runCtx,
-	}, nil
+		intents:              NewIntents(),
+	}
+
+	if err := r.setupTriggerCallbacks(intentChan); err != nil {
+		return nil, errors.Wrapf(err, "setting up trigger callbacks")
+	}
+
+	return r, nil
+}
+
+func (r *SkaffoldRunner) setupTriggerCallbacks(c chan bool) error {
+	if err := r.setupTriggerCallback("build", c); err != nil {
+		return err
+	}
+	if err := r.setupTriggerCallback("sync", c); err != nil {
+		return err
+	}
+	if err := r.setupTriggerCallback("deploy", c); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *SkaffoldRunner) setupTriggerCallback(triggerName string, c chan bool) error {
+	var (
+		intentSet      func(bool)
+		trigger        string
+		resetCallback  *func()
+		serverCallback func(func())
+	)
+
+	switch triggerName {
+	case "build":
+		intentSet = r.intents.setBuild
+		trigger = r.runCtx.Opts.BuildTrigger
+		serverCallback = server.SetBuildCallback
+		resetCallback = &r.intents.resetBuild
+	case "sync":
+		intentSet = r.intents.setSync
+		trigger = r.runCtx.Opts.SyncTrigger
+		serverCallback = server.SetSyncCallback
+		resetCallback = &r.intents.resetSync
+	case "deploy":
+		intentSet = r.intents.setDeploy
+		trigger = r.runCtx.Opts.DeployTrigger
+		serverCallback = server.SetDeployCallback
+		resetCallback = &r.intents.resetDeploy
+	default:
+		return fmt.Errorf("unsupported trigger type when setting callbacks: %s", triggerName)
+	}
+
+	intentSet(true)
+
+	if trigger == "api" || r.runCtx.Opts.APIMode {
+		intentSet(false) // set the initial value of the intent to false
+		// give the server a callback to set the intent value when a user request is received
+		serverCallback(func() {
+			logrus.Debugf("%s intent received, calling back to runner", triggerName)
+			c <- true
+			intentSet(true)
+		})
+		// expose the callback to reset the value inside the intent struct
+		*resetCallback = func() {
```

Comment:
ohhh, I see...hmm..I wonder if this ever changes to anything else? Why can't this definition just simply live in the intent struct?